src/spider.js

```diff
@@ -0,0 +1,268 @@
+'use strict'
+
+const http = require('http')
+const https = require('https')
+const LE = require('greenlock').LE
+const leChallengeFs = require('le-challenge-fs')
+const leAcmeCore = require('le-acme-core')
+const leStoreBot = require('le-store-certbot')
+const WebSocket = require('ws')
+const BtpPacket = require('btp-packet')
+BtpPacket.isRequest = (type) => {
+  return ([BtpPacket.TYPE_MESSAGE, BtpPacket.TYPE_PREPARE, BtpPacket.TYPE_FULFILL, BtpPacket.TYPE_REJECT].indexOf(type) !== -1)
+}
+
+const BTP_VERSION = BtpPacket.BTP_VERSION_1
+const WELCOME_TEXT = 'This is a BTP server, please upgrade to WebSockets.'
+const LE_ROOT = '~/letsencrypt'
+const HTTP_REDIRECT_PORT = 80
+const HTTPS_PORT = 443
+
+const PEER_TYPE_DOWNSTREAM = 'downstream'
+const PEER_TYPE_UPSTREAM = 'upstream'
+
+const CONNECT_RETRY_INTERVAL = 1000
+const SEND_RETRY_INTERVAL = 100
+
+// This function starts a TLS webserver on HTTPS_PORT, with on-the-fly LetsEncrypt cert registration.
```

Comment:
What's the point of having this get a Let's Encrypt cert for a test server? Why not just use something like localtunnel if it's only for development? Seems like overkill

Reply:
The idea is you could use this to build demos and ILP-enabled services. If you want to use it for trying things on localhost, use the `listen` instead of the `tls` field in the config. You can also use `listen` + localtunnel if that's what you want, or use `listen` and then add a letsencrypt-nginx proxy, like ilp-kit-docker-compose does. BTP Cat just offers the two options, it's up to you what you want to combine them with.

Reply:
For demos you'd probably want the simplest thing possible, so I don't think you'd run it on a remote server.

For building ILP-enabled services, what benefit does this module provide? I thought it's mostly for debugging.

Reply:
Assuming that ILP-enabled service has at least one BTP peer, and that it's written in JS, you cannot build it without at least a WebSocket server or a WebSocket client. So if you don't use BtpSpider, you would use websocket/ws and greenlock directly, basically writing the same code, but inside your project. So the benefit is that you don't have to write that code yourself :)

BtpCat is mainly for debugging, although you can also use it for your server logs.

BtpFrog and BtpSpider can of course be used for debugging a BTP-enabled service, but also for building one, or for building a proxy to between your non-JS code and some ledger.

Reply:
I don't think I fully understand your point about "I thought it's mostly for debugging."

Can you more specifically point to lines of code in this PR that you think are not useful when running in production?

Reply:
> if you don't use BtpSpider, you would use websocket/ws and greenlock directly

This is what the plugin is for, no? https://github.com/interledgerjs/ilp-plugin-payment-channel-framework/pull/21 will make it support BTP, including websockets, and @justmoon + @sharafian mentioned adding a websocket server in there as well.

The frog concept makes sense because it turns an internal ledger plugin into an external ledger adapter. The spider part I'm still a little confused about.

(Overall, I'm not sure I'm a fan of the animal naming scheme. I keep having to go look at the readme to remind myself which one is which, and I think people that aren't familiar with the project will find it all the more confusing)

Reply:
yeah, I think the plugin should take a WebSocket object, so it can be agnostic to whether it's a client or a server. That would make it supereasy to use `ilp` + `ilp-plugin-virtual` + BtpSpider to create a fully-functional sender or receiver, whether it be a peered as a client, or as a server.

We can also duplicate the code, that's in itself not very harmful, as it's not that many lines of code. But both Amundsen and Interfaucet use Spider, and Amundsen additionally uses Frog.
You could argue Amundsen should use ilp-connector, but then it still needs some sort of middleware around that. For Interfaucet, that's a good example of a service that uses Spider and almost nothing else.

I'm fine with renaming them, but I don think we should offer tools like this in the BTP toolkit.

Reply:
Why not just include the code for opening the websocket in the plugin and avoid having an extra module?

Reply:
because you don't always want to use a connector plugin. For instance, for a Frog, a logging proxy, but also the Amundsen and Interledger servers. They use BTP directly.

Reply:
Why? What functionality do you need that the plugin doesn't expose? I think it would be better to reuse a single implementation of BTP across multiple components instead of reimplementing it in a bunch of different places, no?

Reply:
So do you feel there should be no WebSocket-related code in the BTP toolbox at all? I don't see how we could create a BTP ledger adapter without a WebSocket. Including it from ilp-plugin-paychan-framework would feel superweird.

Or would it be acceptable if I remove the LetsEncrypt functionality, so that it becomes more of a localhost-only tool?

About the renaming, maybe 'frog' can be called 'adapter', since that's a term already used in the java-interledger community, spider could maybe be split into 'server' and 'client', and Cat could be renamed to Logger or PrettyPrint or something?

Reply:
The main thing is that I don't think we need multiple implementations of BTP in Javascript. If the `ilp-plugin-paychan-framework` includes too much other stuff (like handling trustline/plugin state), maybe BTP should be refactored out into a standalone library that the plugin can then use along with these other modules.

- Frog: It would make sense to me for the ledger adapter tool to include 2 plugins: one for BTP and one for whatever else you're using, and then it would just hook them up. This is basically just a 1:1 connector (also pretty close to @justmoon's connector-shard), so it might be better to think of it like that.
- Spider: What is this trying to accomplish above and beyond what a plugin that implements BTP would do?
- Cat: BTPLogger or BTPPrettyPrint sounds good to me

Reply:
ok, I'll rename BtpCat to BtpLogger, merge that, and close the other two PRs.